src/boot/sign.c

```diff
@@ -0,0 +1,549 @@
+
+        const IMAGE_DATA_DIRECTORY *certificate_table;
+        certificate_table = pe_header_get_data_directory(pe_header, IMAGE_DATA_DIRECTORY_INDEX_CERTIFICATION_TABLE);
+        if (!certificate_table)
+                return log_error_errno(SYNTHETIC_ERRNO(EBADMSG), "File lacks certificate table.");
+
+        off_t end = lseek(fd, 0, SEEK_END);
+        if (end < 0)
+                return log_error_errno(errno, "Failed to seek to end of file: %m");
+
+        if (end % 8 != 0) {
+                if (certificate_table->VirtualAddress != 0)
+                        return log_error_errno(SYNTHETIC_ERRNO(EBADMSG), "Certificate table is not aligned to 8 bytes");
+
+                r = loop_write(fd, (const uint8_t[8]) {}, 8 - (end % 8));
+                if (r < 0)
+                        return log_error_errno(r, "Failed to write zero padding: %m");
+        }
+
+        uint32_t certsz = offsetof(WIN_CERTIFICATE, bCertificate) + sigsz;
+        r = loop_write(fd,
+                       &(WIN_CERTIFICATE) {
+                                .wRevision = htole16(0x200),
+                                .wCertificateType = htole16(0x0002), /* PKCS7 signedData */
+                                .dwLength = htole32(ROUND_UP(certsz, 8)),
+                       },
+                       sizeof(WIN_CERTIFICATE));
+        if (r < 0)
+                return log_error_errno(r, "Failed to write certificate header: %m");
+
+        r = loop_write(fd, sig, sigsz);
+        if (r < 0)
+                return log_error_errno(r, "Failed to append signature: %m");
+
+        if (certsz % 8 != 0) {
+                r = loop_write(fd, (const uint8_t[8]) {}, 8 - (certsz % 8));
+                if (r < 0)
+                        return log_error_errno(r, "Failed to write zero padding: %m");
+        }
+
+        ssize_t n = pwrite(fd,
+                           &(IMAGE_DATA_DIRECTORY) {
+                                .VirtualAddress = certificate_table->VirtualAddress ?: htole32(ROUND_UP(end, 8)),
+                                .Size = htole32(le32toh(certificate_table->Size) + ROUND_UP(certsz, 8)),
+                           },
+                           sizeof(IMAGE_DATA_DIRECTORY),
+                           le32toh(dos_header->e_lfanew) + PE_HEADER_OPTIONAL_FIELD_OFFSET(pe_header, DataDirectory[IMAGE_DATA_DIRECTORY_INDEX_CERTIFICATION_TABLE]));
+        if (n < 0)
+                return log_error_errno(errno, "Failed to update PE certificate table: %m");
+        if ((size_t) n != sizeof(IMAGE_DATA_DIRECTORY))
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Short write while updating PE certificate table.");
+
+        if (lseek(fd, SEEK_SET, 0) < 0)
+                return log_error_errno(errno, "Failed to reset to beginning of file: %m");
+
+        uint32_t checksum;
+        r = pe_checksum(fd, &checksum);
+        if (r < 0)
+                return log_error_errno(r, "Failed to calculate PE file checksum: %m");
```

Comment:
tbh I am not too much of a fan that there's a mixture of pread/pwrite and lseek/read/write used on the same fd. Stick to one please, at least for each direction. Given that pe_hash() and pe_load_headers() all use pread() I think the read side should stick to pread() entirely. And the write side probably too, but I care more about that each direction sticks to one way of doing things than that all of them are doing sames with the same family of calls.

Reply:
@poettering We only use `pread()` already for reading?

Doing unconditional `pwrite()` means writing a version of `loop_write()` built on `pwrite()` which I'd prefer not to do.

Reply:
no, not needed. loop_write() is only necessary for fds such as sockets and pipes and char devices, where short writes are a thing.

But this is for a regular file here, guaranteed, as pread()/pwrite() wouldn#t work on non-regular files anyway. But on regular files read()/write()/pread()/pwrite() already give you the guarantee that they fully succeed, hence no need for the looping.

Reply:
pa_checksum() uses plain read() instead of pread() btw. Hence you are mixing the two flavours on both read and write.

Reply:
(i mean, there are tape drives which are character devices that also implement seeking, as weird outliers of fs that do short read/write but also do seeking. But I am quite sure they do not support pread()/pwrite() either...)

to be fully robust you could of course add fd_verify_regular() immediately after opening the file, then you have your guarantee that pread()/pwrite() will work and never doo short reads/writes

Reply:
i mean, implementing a looping pread()/pwrite() wrapper wouldn't hurt I guess, but I don't think it's necessary. it would be more in the "paranoia" territory I guess. which is a fine territory to be in, but also doesn#t matter